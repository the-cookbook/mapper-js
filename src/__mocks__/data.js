export default [
  {
    _id: '5f298c3a9de3ab8556c084db',
    index: 0,
    guid: '1ee5f65e-d6d3-4cb4-b529-61ef47a152a7',
    isActive: true,
    balance: '$1,273.07',
    picture: 'http://placehold.it/32x32',
    age: 32,
    eyeColor: 'blue',
    name: 'Gay Landry',
    gender: 'female',
    company: 'XANIDE',
    email: 'gaylandry@xanide.com',
    phone: '+1 (955) 541-2541',
    address: '740 Debevoise Avenue, Silkworth, Oregon, 2792',
    about:
      'Cillum est cillum incididunt laborum Lorem ad sit et enim exercitation laboris pariatur. Sit labore occaecat et reprehenderit labore dolore exercitation cupidatat labore esse eu eiusmod. Amet ipsum qui ex id.\r\n',
    registered: '2014-11-05T04:54:40 -01:00',
    latitude: -21.237512,
    longitude: -79.117184,
    tags: ['amet', 'labore', 'irure', 'in', 'proident', 'qui', 'nulla'],
    friends: [
      {
        id: 0,
        name: 'Courtney Medina',
      },
      {
        id: 1,
        name: 'Hodges Avery',
      },
      {
        id: 2,
        name: 'Blanchard Hyde',
      },
    ],
    greeting: 'Hello, Gay Landry! You have 2 unread messages.',
    favoriteFruit: 'banana',
  },
  {
    _id: '5f298c3a79466b7d76789046',
    index: 1,
    guid: 'acc54e37-8bbc-49a4-ba41-3a57aa415f4a',
    isActive: false,
    balance: '$3,134.54',
    picture: 'http://placehold.it/32x32',
    age: 24,
    eyeColor: 'blue',
    name: 'Harris Bryant',
    gender: 'male',
    company: 'KINDALOO',
    email: 'harrisbryant@kindaloo.com',
    phone: '+1 (881) 504-2864',
    address: '406 Macdougal Street, Katonah, Arkansas, 2252',
    about:
      'Dolore nostrud mollit esse sint sint nostrud elit reprehenderit. Nulla mollit in ullamco quis est. Nulla adipisicing ad deserunt veniam consectetur nostrud irure aute. Et do veniam non esse irure sunt ut aliqua anim pariatur ad velit irure culpa. Adipisicing non veniam anim eu laborum. Excepteur deserunt enim consequat consequat laboris officia excepteur dolore.\r\n',
    registered: '2018-12-07T01:13:53 -01:00',
    latitude: 21.020802,
    longitude: 162.663481,
    tags: ['consequat', 'culpa', 'pariatur', 'incididunt', 'sunt', 'enim', 'dolor'],
    friends: [
      {
        id: 0,
        name: 'Ruby Conley',
      },
      {
        id: 1,
        name: 'Morgan Ramos',
      },
      {
        id: 2,
        name: 'Jeanne Woodward',
      },
    ],
    greeting: 'Hello, Harris Bryant! You have 7 unread messages.',
    favoriteFruit: 'banana',
  },
  {
    _id: '5f298c3abbbc5c3e156cee62',
    index: 2,
    guid: '0b1b7bb7-4ac1-4c75-8fbe-bdc1ca8854e0',
    isActive: true,
    balance: '$1,871.29',
    picture: 'http://placehold.it/32x32',
    age: 32,
    eyeColor: 'blue',
    name: 'Rosario Delacruz',
    gender: 'female',
    company: 'ENTOGROK',
    email: 'rosariodelacruz@entogrok.com',
    phone: '+1 (813) 508-3145',
    address: '848 College Place, Glenshaw, New York, 1218',
    about:
      'Est tempor et id officia occaecat ad. In fugiat sint do velit in veniam laboris laboris consectetur labore. Eu qui velit sunt laborum. Quis laboris ut enim nisi irure id dolor dolor commodo aliqua non in. Cillum aliqua adipisicing ex nostrud eu ipsum in dolore commodo aliquip ut.\r\n',
    registered: '2020-04-29T04:59:36 -02:00',
    latitude: 62.267437,
    longitude: 4.248991,
    tags: ['ut', 'irure', 'nulla', 'esse', 'labore', 'occaecat', 'nisi'],
    friends: [
      {
        id: 0,
        name: 'Lesley Hebert',
      },
      {
        id: 1,
        name: 'Ramos Booth',
      },
      {
        id: 2,
        name: 'Decker Hubbard',
      },
    ],
    greeting: 'Hello, Rosario Delacruz! You have 9 unread messages.',
    favoriteFruit: 'strawberry',
  },
  {
    _id: '5f298c3abad7589e14b8ea22',
    index: 3,
    guid: '4297fcc6-ee7c-4ff5-ae62-768977358359',
    isActive: false,
    balance: '$3,377.96',
    picture: 'http://placehold.it/32x32',
    age: 36,
    eyeColor: 'blue',
    name: 'Lorene Hays',
    gender: 'female',
    company: 'ECOSYS',
    email: 'lorenehays@ecosys.com',
    phone: '+1 (852) 436-2074',
    address: '683 Cornelia Street, Edgar, Pennsylvania, 201',
    about:
      'Magna Lorem magna nulla ipsum aute in velit cupidatat occaecat. Est dolor cillum ad aliquip in reprehenderit enim. Voluptate dolore minim voluptate nisi nisi ipsum duis nisi.\r\n',
    registered: '2014-07-02T04:22:37 -02:00',
    latitude: -14.832106,
    longitude: 97.156766,
    tags: ['excepteur', 'dolor', 'voluptate', 'ea', 'ea', 'esse', 'sit'],
    friends: [
      {
        id: 0,
        name: 'Lillie Holloway',
      },
      {
        id: 1,
        name: 'Jenna Chandler',
      },
      {
        id: 2,
        name: 'Nguyen Shannon',
      },
    ],
    greeting: 'Hello, Lorene Hays! You have 1 unread messages.',
    favoriteFruit: 'apple',
  },
  {
    _id: '5f298c3ae53e4116a9df500f',
    index: 4,
    guid: 'ef899c53-cbee-4ee0-be36-a84ed418a33b',
    isActive: true,
    balance: '$1,697.40',
    picture: 'http://placehold.it/32x32',
    age: 38,
    eyeColor: 'green',
    name: 'Fern Holcomb',
    gender: 'female',
    company: 'BUNGA',
    email: 'fernholcomb@bunga.com',
    phone: '+1 (886) 586-3726',
    address: '565 Hinsdale Street, Chicopee, Federated States Of Micronesia, 5333',
    about:
      'Id ea cillum cupidatat ullamco exercitation commodo elit sint id dolore velit officia. Non sunt dolor amet aute voluptate ipsum eiusmod. Amet consequat voluptate do amet culpa ea.\r\n',
    registered: '2017-10-20T12:02:15 -02:00',
    latitude: 36.949756,
    longitude: -63.13798,
    tags: ['ea', 'ipsum', 'dolore', 'elit', 'exercitation', 'ullamco', 'dolor'],
    friends: [
      {
        id: 0,
        name: 'Latisha Cochran',
      },
      {
        id: 1,
        name: 'Ramsey Duke',
      },
      {
        id: 2,
        name: 'Terrell Dickerson',
      },
    ],
    greeting: 'Hello, Fern Holcomb! You have 8 unread messages.',
    favoriteFruit: 'apple',
  },
  {
    _id: '5f298c3a7c82300acb59f925',
    index: 5,
    guid: 'acf11664-0349-4967-962d-ebf80bbfa498',
    isActive: false,
    balance: '$2,188.03',
    picture: 'http://placehold.it/32x32',
    age: 34,
    eyeColor: 'blue',
    name: 'Sadie Sloan',
    gender: 'female',
    company: 'CAXT',
    email: 'sadiesloan@caxt.com',
    phone: '+1 (996) 460-2068',
    address: '105 Lafayette Walk, Galesville, Texas, 7311',
    about:
      'Dolor nisi culpa magna et minim cillum reprehenderit incididunt elit qui excepteur. In ad dolor culpa occaecat id voluptate quis amet consequat. Sit elit excepteur velit aliquip id quis deserunt officia duis. Tempor sunt do quis velit elit in commodo consectetur officia sunt ipsum.\r\n',
    registered: '2016-12-11T09:46:02 -01:00',
    latitude: 45.161667,
    longitude: 72.818628,
    tags: ['velit', 'adipisicing', 'amet', 'anim', 'labore', 'tempor', 'ea'],
    friends: [
      {
        id: 0,
        name: 'Jennings Mays',
      },
      {
        id: 1,
        name: 'Mendoza Stone',
      },
      {
        id: 2,
        name: 'Yang Espinoza',
      },
    ],
    greeting: 'Hello, Sadie Sloan! You have 2 unread messages.',
    favoriteFruit: 'apple',
  },
  {
    _id: '5f298c3aff8283a5ca3aa82d',
    index: 6,
    guid: '36b9df6a-b846-4940-80cc-c9d163d2128c',
    isActive: false,
    balance: '$3,222.06',
    picture: 'http://placehold.it/32x32',
    age: 38,
    eyeColor: 'green',
    name: 'Doris Mcfarland',
    gender: 'female',
    company: 'MAGNINA',
    email: 'dorismcfarland@magnina.com',
    phone: '+1 (835) 486-2425',
    address: '367 Drew Street, Kiskimere, Arizona, 3542',
    about:
      'Aute ut sit aliquip mollit irure. Tempor officia ut laboris anim minim sit ipsum quis exercitation anim proident ipsum voluptate. Nisi anim ad consequat velit culpa deserunt dolore sit incididunt. Sint aliquip enim irure laboris duis. Officia aute irure ut nisi aute elit mollit culpa eiusmod consequat laboris adipisicing.\r\n',
    registered: '2014-04-25T04:58:14 -02:00',
    latitude: 21.082733,
    longitude: -89.945592,
    tags: ['nulla', 'et', 'pariatur', 'nisi', 'aliquip', 'irure', 'deserunt'],
    friends: [
      {
        id: 0,
        name: 'Marcy Lucas',
      },
      {
        id: 1,
        name: 'Christensen Schwartz',
      },
      {
        id: 2,
        name: 'Amanda Valencia',
      },
    ],
    greeting: 'Hello, Doris Mcfarland! You have 2 unread messages.',
    favoriteFruit: 'apple',
  },
];
