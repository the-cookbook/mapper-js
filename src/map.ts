import { pick } from '@cookbook/dot-notation';

import { defaultOptions, Options } from './options';
import toArray from './utils/to-array';
import is from './utils/is';
import typeOf from './utils/type-of';

type TransformCallback = (...args: unknown[]) => unknown | unknown[];

type MapMethods<T = unknown> = {
  transform(callback: TransformCallback): MapMethods<T>;
  readonly value: T extends [] ? T[] : T;
};

type Map = <T = unknown>(keys: string | string[], options?: Options) => MapMethods<T>;

const suppress = Symbol('map-suppress');

const map = <T = unknown>(source: Record<string, unknown | unknown[]>, mapperOptions: Options = {}): Map => {
  if (!is.object(source)) {
    throw new TypeError(`Instance of "source" must be an object, but instead got "${typeOf(source)}"`);
  }

  return <T = unknown>(keys: string | string[], options: Options = {}) => {
    const OPTIONS = { ...defaultOptions, ...mapperOptions, ...options };

    let result: unknown | unknown[] = is.array(keys) ? keys.map((key) => pick(source, key)) : pick(source, keys);

    return {
      transform(callback: TransformCallback) {
        result = is.array(keys) ? callback(...(<unknown[]>result)) : callback(result);

        return this;
      },
      get value(): T extends [] ? T[] : T {
        const { omitNullUndefined, omitStrategy } = OPTIONS;

        if ((is.nullOrUndefined(result) && omitNullUndefined) || (omitStrategy && omitStrategy(result))) {
          return (suppress as unknown) as T extends [] ? T[] : T;
        }

        return result as T extends [] ? T[] : T;
      },
    };
  };
};

map.omitEntries = (
  entries: Record<string, unknown | unknown[]> | Record<string, unknown | unknown[]>[],
): Record<string, unknown | unknown[]> | Record<string, unknown | unknown[]>[] => {
  const result = toArray(entries).map((entry) => {
    return Object.keys(entry).reduce((acc, key) => {
      if (entry[key] !== suppress) {
        acc[key] = entry[key];
      }

      return acc;
    }, {} as Record<string, unknown | unknown[]>);
  });

  return is.array(entries) ? result : result[0];
};

map.suppress = suppress;

export type { Map, MapMethods, TransformCallback };
export default map;
