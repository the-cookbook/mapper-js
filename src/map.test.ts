/* eslint-disable @typescript-eslint/no-explicit-any, jest/no-mocks-import */
import map from './map';
import data from './__mocks__/data';

describe('map()', () => {
  it('should match snapshot', () => {
    expect(map).toMatchSnapshot();
  });

  describe('exceptions', () => {
    it('should throw an error when source is not an object', () => {
      [true, '', 0, null, undefined, [0]].forEach((source) => {
        expect(() => map(source as any)('person.address[0]').value).toThrow(TypeError);
      });
    });
  });

  describe('default', () => {
    it('should map single keys', () => {
      expect(map(data[0])('address[0]').value).toEqual(data[0].address[0]);
      expect(map(data[0])('email').value).toEqual(data[0].email);
      expect(map(data[0])('name').transform((name: string) => name.replace(/\s/gm, '-')).value).toEqual(
        data[0].name.replace(/\s/gm, '-'),
      );
    });

    it('should map multiple keys', () => {
      expect(
        map(data[2])(['latitude', 'longitude']).transform((latitude: number, longitude: number) => [
          latitude,
          longitude,
        ]).value,
      ).toEqual([data[2].latitude, data[2].longitude]);

      expect(
        map(data[2])(['phone', 'email']).transform((phone: number, email: number) => ({
          phone,
          email,
        })).value,
      ).toEqual({
        phone: data[2].phone,
        email: data[2].email,
      });
    });
  });

  describe('options', () => {
    describe('omitNullUndefined', () => {
      it('should omit "null" and "undefined" entries', () => {
        expect(map(data[0])('closeFriends', { omitNullUndefined: true }).value).toEqual(map.suppress);
        expect(map(data[0])('non-existing', { omitNullUndefined: true }).value).toEqual(map.suppress);
      });
    });

    describe('omitStrategy', () => {
      it('should omit entries through custom strategy', () => {
        const omitStrategy = (value: unknown | unknown[]): boolean => Array.isArray(value);

        expect(map(data[0])('friends', { omitStrategy }).value).toEqual(map.suppress);
        expect(map(data[0])('tags', { omitStrategy }).value).toEqual(map.suppress);
      });
    });
  });
});
/* eslint-enable @typescript-eslint/no-explicit-any, jest/no-mocks-import */
