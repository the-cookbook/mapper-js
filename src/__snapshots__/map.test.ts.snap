// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`map() should match snapshot 1`] = `[Function]`;
